### ISSUE TITLE
`g_csrf_token` cookie is not received on a different subdomain

### ISSUE THREAD

**Issue by jotaen** [googleapis/google-auth-library-nodejs#1241]
# Problem
I don’t receive the `g_csrf_token` cookie on my server.

# What I did
- I have placed the “Continue with Google” login button on my login page, under a URL like so `https://example.org/login.html` (no subdomain!).
- My login API endpoint is `https://auth.example.org/google` (different subdomain!), which I configured in the [`data_login-uri`](https://developers.google.com/identity/gsi/web/reference/html-reference#data-login_uri) attribute of the button.

If I click on the “Continue with Google” button, the following happens:

1. The [JS-library](https://accounts.google.com/gsi/client) sets a cookie called `g_csrf_token`.
2. I’m externally redirected to the Google account portal to select my account to login with.
3. After selecting my account, I’m redirected to the configured login URI `https://auth.example.org/google`.

The problem is in the last step, because the browser does not include the `g_csrf_token` cookie in that request. The reason is that it’s a `hostOnly` cookie, which will only be included in requests to the **exact** same domain where it was set on. The cookie was set on `example.org`, but the request goes to `auth.example.org`.

# Possible solution
Provide a way to set the `domain` attribute of the cookie explicitly, e.g. via `data-cookie_domain`. In this case the cookie won’t be marked as `hostOnly` and will be included in requests across all subdomains.

---

Btw., I’ve also tried to set [`data-state_cookie_domain`](https://developers.google.com/identity/gsi/web/reference/html-reference#data-state_cookie_domain), but that didn’t solve the issue. I’m also not using the “One Tap” flow, but the “Redirect” one.

For reference, my button config looks like so:

```html
<div
  id='g_id_onload'
  data-client_id='1234567890-xyzxyzxyzxyzxyz.apps.googleusercontent.com'
  data-context='signin'
  data-ux_mode='redirect'
  data-login_uri='https://auth.example.org/google'
  data-auto_prompt='false'>
</div>
```

I’m aware that this is probably not the right place to open this issue, but I’m not sure where else to bring this up. I’m using the NodeJS library in the backend, though, so I thought I can start to ask here. If there is a better place, please let me know. Thanks!

**Comment by sofisl** [comment]
Hi @jotaen,

Thanks for submitting! After doing some investigating, I think your issue is really with this [feature](https://developers.google.com/identity/sign-in/web/sign-in), and you can submit your bug [here](https://stackoverflow.com/questions/tagged/google-signin) and/or [here](https://github.com/googlesamples/google-services).

**Comment by jotaen** [comment]
I’m not sure whether StackOverflow or the GoogleSamples repo are the right places to ask either. As you mentioned, I also think the issue is with the feature itself and with the JS library, unfortunately I have no idea who to reach out to.

**Comment by spacesEdward** [comment]
I ran into the same problem and didn't find anything related in SO, so I've asked a question here: https://stackoverflow.com/questions/70195261/g-csrf-token-not-passed-as-cookie-to-subdomain

I will update this comment if I get a response that resolves my issue.

**Comment by jotaen** [comment]
Just for the records, I didn’t file a question on SO or the samples repo. I don’t expect anything to come out of it, since the issue is with the library itself and not with its usage, so I personally think that only the developers of the library would be able to help out. But there doesn’t seem to be any way to reach them.

I ended up adding a hack into my Proxy server that internally forwards the callback request from Google to my API backend (the one that actually runs on the subdomain). This also means I had to loosen my session cookie’s policy and scope the cookie to the entire domain (`*.domain.com`) instead of just the API subdomain. If altering the scope is not an option, another potential approach might be an external token redirect from the main domain to the subdomain, so that the cookie can still be issued (and thus scoped) from/to the subdomain. That still would require the proxy hack, however.
